Moves cleaning to an atom proc to afterattack signal

JohnFulpWillard opened the pull request:
## About The Pull Request

* Cleaning off window blood is now handled by the Window itself
* Adds a pre_clean_callback to replace the should_clean arg (mostly), also adds a DO_NOT_CLEAN define that will prevent cleaning entirely, to remove the need to have checks in individual afterattacks and instead do it in the callback (primary example is mop, which shouldn't clean if it has no reagents or is attacking a bucket/janicart)
* Removes the atom cleaning proc, replaces it with a signal for afterattack on a cleaning tool, or bot's afterattack.

## Why It's Good For The Game

Atom procs like these are against what signals should be, hopefully this will be better.
I also noticed that rags check if the rag is in user to work, it seems like a rather useless check? Is it there for some Telekinesis thing or were rags meant to only clean things on the user (so the bartender can wipe down glasses)? I'll probably make that change in the future once I find out.

## Changelog

🆑 
code: Players are no longer told about washing off someone's mouth with soap, since any cleaning tool (including soap) already does it, even if they don't target the mouth.
/🆑

JohnFulpWillard committed:
Hopefully fixes Cleaning

* Cleaning off window blood is now handled by the Window itself
* Cleanbots now have an advanced mop built into them, I am hoping in the future we can make cleanbots have their mops taken off, but I'm not good at spriting so it isn't a thing yet.
* Adds a pre_clean_callback to replace the should_clean arg (mostly)
* Removes the atom cleaning proc, replaces it with a signal for afterattack on a cleaning tool.

kawoppi commented:
Did you check if cleaning something while you're already cleaning it doesn't remove the bubbles animation before the do_after finishes? The already_cleaning var was used to prevent that but I'm not sure if other changes have made it unnecessary.

JohnFulpWillard commented:
> Did you check if cleaning something while you're already cleaning it doesn't remove the bubbles animation before the ongoing do_after finishes? The already_cleaning var was used to prevent that but I'm not sure if other changes have made it unnecessary.

I removed 1/2 times it was used so just made the other time it was used check the trait instead

kawoppi commented:
> I removed 1/2 times it was used so just made the other time it was used check the trait instead

It's not so much about the var itself but the check before cutting the overlays at the end. I think the issue was that if you clean something and then clean it again before the first one finishes, the second cleaning skips the do_after and cuts the overlays before the first cleaning can finish it's do_after. That check made it so it would only cut the overlays if it had also added those overlays earlier.

JohnFulpWillard commented:
> > I removed 1/2 times it was used so just made the other time it was used check the trait instead
> 
> It's not so much about the var itself but the check before cutting the overlays at the end. I think the issue was that if you clean something and then clean it again before the first one finishes, the second cleaning skips the do_after and cuts the overlays before the first cleaning can finish it's do_after. That check made it so it would only cut the overlays if it had also added those overlays earlier.

The code currently adds a trait but only removes it if they don't have the trait, I thought that was silly so I removed it, from my short testing it didn't seem like a problem but I'll test it more now that you brought it up.

JohnFulpWillard committed:
Fixes and improvements

* adds a new define that allows pre_clean_callback to cancel cleaning entirely, used for mops to check for being wet or attacking buckets and such.
* Fixes cleanbots not properly working, removing their mop and just registering for a signal that bots send instead.
* Changes cleaner component to take proximity into account, that way we don't have to check it on each item's individual afterattacks, that way we don't have to call parent at the end either.
* Improves check_allowed_items a little bit because I hated how it was used for soap.

JohnFulpWillard commented:
Tested this stuff in-game and it works just as expected, however cleanbots were broken. I fixed those and also changed a few other stuff as well.

JohnFulpWillard committed:
Update cleaner.dm

Fikou commented on code/modules/mob/living/simple_animal/bot/cleanbot.dm:
```diff
@@ -267,7 +268,7 @@
 				path = list()
 			else
-				UnarmedAttack(target) //Rather than check at every step of the way, let's check before we do an action, so we can rescan before the other bot.
+				UnarmedAttack(target, TRUE) //Rather than check at every step of the way, let's check before we do an action, so we can rescan before the other bot.
```
named arguments

Fikou commented on code/modules/mob/living/simple_animal/bot/cleanbot.dm:
```diff
@@ -239,9 +240,9 @@
+			UnarmedAttack(victim, TRUE) // Acid spray
 		if(isopenturf(loc) && prob(15)) // Wets floors and spawns foam randomly
-			UnarmedAttack(src)
+			UnarmedAttack(src, TRUE)
```
here too

Fikou commented on code/modules/mob/living/simple_animal/bot/cleanbot.dm:
```diff
@@ -239,9 +240,9 @@
 		var/mob/living/carbon/victim = locate(/mob/living/carbon) in loc
 		if(victim && victim == target)
-			UnarmedAttack(victim) // Acid spray
+			UnarmedAttack(victim, TRUE) // Acid spray
```
here too

Fikou commented on code/modules/mob/living/simple_animal/bot/cleanbot.dm:
```diff
@@ -121,7 +122,7 @@
 	. = ..()
 	if(!weapon)
-		return
+		return .
```
dont do that

Fikou replied:
i hate mothblocks

Fikou commented on code/modules/mob/living/simple_animal/bot/cleanbot.dm:
```diff
@@ -333,12 +334,10 @@
 		update_icon_state()
-		var/turf/turf_to_clean = get_turf(attack_target)
-		start_cleaning(src, turf_to_clean, src)
+		..()
```
did you mean to not set return value anymore

JohnFulpWillard replied:
It wasn't necessary to have

Fikou commented on code/modules/reagents/reagent_containers/misc.dm:
```diff
@@ -131,29 +131,31 @@
 		return
-	if(iscarbon(A) && reagents?.total_volume)
-		var/mob/living/carbon/C = A
+	if(iscarbon(target) && reagents?.total_volume)
```
early return here please

JohnFulpWillard committed:
reviews
